File: backend/microservice1/src/main/java/tn/esprit/microservice1/services/SportService.java
package tn.esprit.microservice1.services;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import tn.esprit.microservice1.entities.Sport;
import tn.esprit.microservice1.entities.SalleDeSport;
import tn.esprit.microservice1.repositories.SportRepository;
import tn.esprit.microservice1.repositories.SalleDeSportRepository;

import java.util.List;
import java.util.Set;

@Service
public class SportService {
    
    private final SportRepository sportRepository;
    private final SalleDeSportRepository salleDeSportRepository;
    
    @Autowired
    public SportService(SportRepository sportRepository, SalleDeSportRepository salleDeSportRepository) {
        this.sportRepository = sportRepository;
        this.salleDeSportRepository = salleDeSportRepository;
    }
    
    public List<Sport> getAllSports() {
        return sportRepository.findAll();
    }
    
    public Sport getSportById(Integer id) {
        return sportRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Sport not found with id: " + id));
    }
    
    public Sport createSport(Sport sport) {
        return sportRepository.save(sport);
    }
    
    public Sport updateSport(Integer id, Sport sport) {
        if (!sportRepository.existsById(id)) {
            throw new RuntimeException("Sport not found with id: " + id);
        }
        sport.setCodeSport(id);
        return sportRepository.save(sport);
    }
    
    public void deleteSport(Integer id) {
        if (!sportRepository.existsById(id)) {
            throw new RuntimeException("Sport not found with id: " + id);
        }
        sportRepository.deleteById(id);
    }
    
    // Relationship management methods
    public void assignSportToSalle(Integer sportId, Integer salleId) {
        Sport sport = getSportById(sportId);
        SalleDeSport salle = salleDeSportRepository.findById(salleId)
                .orElseThrow(() -> new RuntimeException("Salle not found with id: " + salleId));
        
        salle.addSport(sport);
        salleDeSportRepository.save(salle);
    }
    
    public void removeSportFromSalle(Integer sportId, Integer salleId) {
        Sport sport = getSportById(sportId);
        SalleDeSport salle = salleDeSportRepository.findById(salleId)
                .orElseThrow(() -> new RuntimeException("Salle not found with id: " + salleId));
        
        salle.removeSport(sport);
        salleDeSportRepository.save(salle);
    }
    
    public Set<SalleDeSport> getSallesBySportId(Integer sportId) {
        Sport sport = getSportById(sportId);
        return sport.getSalles();
    }
} 

File: frontend/backoffice/src/app/microservice1/services/sport.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Sport } from '../models/sport.model';
import { SalleDeSport } from '../models/salle-de-sport.model';
import { environment } from '../../../environments/environment';

@Injectable({
  providedIn: 'root'
})
export class SportService {
  private apiUrl = `${environment.apiUrl}/mic1/sports`;

  constructor(private http: HttpClient) { }

  getAllSports(): Observable<Sport[]> {
    return this.http.get<Sport[]>(this.apiUrl);
  }

  getSportById(id: number): Observable<Sport> {
    return this.http.get<Sport>(`${this.apiUrl}/${id}`);
  }

  createSport(sport: Sport): Observable<Sport> {
    return this.http.post<Sport>(this.apiUrl, sport);
  }

  updateSport(id: number, sport: Sport): Observable<Sport> {
    return this.http.put<Sport>(`${this.apiUrl}/${id}`, sport);
  }

  deleteSport(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  // Relationship management
  assignSportToSalle(sportId: number, salleId: number): Observable<void> {
    return this.http.post<void>(`${this.apiUrl}/${sportId}/salles/${salleId}`, {});
  }

  removeSportFromSalle(sportId: number, salleId: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${sportId}/salles/${salleId}`);
  }

  getSallesBySportId(sportId: number): Observable<SalleDeSport[]> {
    return this.http.get<SalleDeSport[]>(`${this.apiUrl}/${sportId}/salles`);
  }
} 